overlord/state/taskrunner.go

```diff
@@ -42,93 +40,107 @@ type TaskRunner struct {
-// tasks of the given kind.
-func (r *TaskRunner) AddHandler(kind string, fn HandlerFunc) {
+// AddHandler registers the functions to concurrently call for doing and
+// undoing tasks of the given kind. The undo handler may be nil.
+func (r *TaskRunner) AddHandler(kind string, do, undo HandlerFunc) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
-	r.handlers[kind] = fn
+	r.handlers[kind] = handlerPair{do, undo}
 }
 
-// propagateError sets all tasks directly and indirectly waiting on t to ErrorStatus.
-func propagateError(task *Task) {
-	mark := append([]*Task(nil), task.HaltTasks()...)
-	for i := 0; i < len(mark); i++ {
-		t := mark[i]
-		t.SetStatus(ErrorStatus)
-		mark = append(mark, t.HaltTasks()...)
+// run must be called with the state lock in place
+func (r *TaskRunner) run(t *Task) {
+	var handler HandlerFunc
+	switch t.Status() {
+	case DoStatus:
+		t.SetStatus(DoingStatus)
+		fallthrough
+	case DoingStatus:
+		handler = r.handlers[t.Kind()].do
+
+	case UndoStatus:
+		t.SetStatus(UndoingStatus)
+		fallthrough
+	case UndoingStatus:
+		handler = r.handlers[t.Kind()].undo
+
+	default:
+		panic("internal error: attempted to run task in status " + t.Status().String())
+	}
+	if handler == nil {
+		panic("internal error: attempted to run task with nil handler for status " + t.Status().String())
 	}
-}
 
-// run must be called with the state lock in place
-func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	tomb := &tomb.Tomb{}
-	id := task.ID()
-	r.tombs[id] = tomb
+	r.tombs[t.ID()] = tomb
 	tomb.Go(func() error {
-		defer func() {
-			r.mu.Lock()
-			defer r.mu.Unlock()
-
-			delete(r.tombs, id)
-		}()
-
-		// capture the error result with tomb.Kill so we can
+		// Capture the error result with tomb.Kill so we can
 		// use tomb.Err uniformily to consider both it or a
 		// overriding previous Kill reason.
-		tomb.Kill(fn(task, tomb))
+		tomb.Kill(handler(t, tomb))
 
+		r.mu.Lock()
+		defer r.mu.Unlock()
 		r.state.Lock()
 		defer r.state.Unlock()
+
+		delete(r.tombs, t.ID())
+
 		switch err := tomb.Err(); err {
 		case Retry:
-			// Do nothing. Handler asked to try again later.
-			// TODO: define how to control retry intervals,
-			// right now things will be retried at the next Ensure
+			// Handler asked to be called again later.
+			// TODO Allow postponing retries past the next Ensure.
+			if t.Status() == AbortStatus {
+				// Would work without it but might take two ensures.
+				r.swapAbort(t)
+			}
 		case nil:
-			task.SetStatus(DoneStatus)
-			if len(task.HaltTasks()) > 0 {
-				// give a chance to taskrunners Ensure to start
-				// the waiting ones
+			var next []*Task
+			switch t.Status() {
+			case DoingStatus:
+				t.SetStatus(DoneStatus)
+				next = t.HaltTasks()
+			case AbortStatus:
+				t.SetStatus(DoneStatus) // Not actually aborted.
+				r.swapAbort(t)
+			case UndoingStatus:
+				t.SetStatus(UndoneStatus)
+				next = t.WaitTasks()
+			}
+			if len(next) > 0 {
 				r.state.EnsureBefore(0)
```

Comment:
because of the change of style to calling Ensure in a loop in the tests, this is not tested anymore, in fact it can be commented out and all tests still pass. In fact I suspect there a couple more code paths in run() that would preferably end in EnsureBefore (switching any task to Abort probably and to Undo so that we get a timely undoing and not a wait of 5 mins for that to start), we need to get back to figure out how to test this before that's even a question though